cmrf_views/src/Plugin/views/query/API4.php

```diff
@@ -0,0 +1,525 @@
+<?php namespace Drupal\cmrf_views\Plugin\views\query;
+
+use Drupal;
+use Drupal\cmrf_core\Call;
+use Drupal\cmrf_core\Core;
+use Drupal\cmrf_views\CMRFViewsResultRow;
+use Drupal\cmrf_views\Entity\CMRFDataset;
+use Drupal\views\Plugin\views\query\QueryPluginBase;
+use Drupal\views\ResultRow;
+use Drupal\views\ViewExecutable;
+use Drupal\views\ViewsData;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * CiviMRF CiviCRM Views query plugin which wraps calls to the API in order to
+ * expose the results to views.
+ *
+ * @ViewsQuery(
+ *   id = "civicrm_api4",
+ *   title = @Translation("CiviMRF CiviCRM APIv4"),
+ *   help = @Translation("Query against the CiviCRM APIv4.")
+ * )
+ */
+class API4 extends QueryPluginBase {
+
+  /**
+   * @var \Drupal\cmrf_core\Core
+   */
+  protected $core;
+
+  /**
+   * @var \Drupal\views\ViewsData
+   */
+  protected $viewsData;
+
+  /**
+   * API constructor.
+   *
+   * @param array                   $configuration
+   * @param                         $plugin_id
+   * @param                         $plugin_definition
+   * @param \Drupal\cmrf_core\Core  $core
+   * @param \Drupal\views\ViewsData $views_data
+   */
+  public function __construct(
+    array $configuration,
+    $plugin_id,
+    $plugin_definition,
+    Core $core,
+    ViewsData $views_data
+  ) {
+    parent::__construct($configuration, $plugin_id, $plugin_definition);
+    $this->core      = $core;
+    $this->viewsData = $views_data;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(
+    ContainerInterface $container,
+    array $configuration,
+    $plugin_id,
+    $plugin_definition
+  ) {
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('cmrf_core.core'),
+      $container->get('views.views_data')
+    );
+  }
+
+  /**
+   * Method to trick views, because it expects an SQL backend.
+   *
+   * @param      $table
+   * @param null $relationship
+   *
+   * @return string
+   */
+  public function ensureTable($table, $relationship = NULL) {
+    return $table;
+  }
+
+  /**
+   * Method to trick views, because it expects an SQL backend.
+   *
+   * @param        $table
+   * @param        $field
+   * @param string $alias
+   * @param array  $params
+   *
+   * @return mixed
+   */
+  public function addField($table, $field, $alias = '', $params = []) {
```

Comment:
I'd suggest to add PHP type hints including return type unless there's a good reason not to do it. (This holds true for the other methods as well.) If `$alias` is type hinted with string `empty($alias)` below could be replaced by `'' === $alias`.